pkg/cache/tas_flavor_snapshot.go

```diff
@@ -379,26 +382,135 @@ func (s *TASFlavorSnapshot) Fits(flavorUsage workload.TASFlavorUsage) bool {
 // the TAS requests in the flavor handled by the snapshot.
 // The simulateEmpty parameter allows to look for the assignment under the
 // assumption that all TAS workloads are preempted.
-func (s *TASFlavorSnapshot) FindTopologyAssignmentsForFlavor(flavorTASRequests FlavorTASRequests, simulateEmpty bool) TASAssignmentsResult {
+func (s *TASFlavorSnapshot) FindTopologyAssignmentsForFlavor(flavorTASRequests FlavorTASRequests, simulateEmpty bool, wl *kueue.Workload) TASAssignmentsResult {
 	result := make(map[kueue.PodSetReference]tasPodSetAssignmentResult)
 	assumedUsage := make(map[utiltas.TopologyDomainID]resources.Requests)
 	for _, tr := range flavorTASRequests {
-		assignment, reason := s.findTopologyAssignment(tr, assumedUsage, simulateEmpty)
-		result[tr.PodSet.Name] = tasPodSetAssignmentResult{TopologyAssignment: assignment, FailureReason: reason}
-		if reason != "" {
-			return result
-		}
-		for _, domain := range assignment.Domains {
-			domainID := utiltas.DomainID(domain.Values)
-			if assumedUsage[domainID] == nil {
-				assumedUsage[domainID] = resources.Requests{}
+		if workload.HasFailedNodeAnnotation(wl) {
+			// In case of looking for replacement TopologyRequest has only
+			// PodSet with node to replace, so we match PodSetAssignment
+			psa := findPSA(wl, tr.PodSet.Name)
+			if psa == nil || psa.TopologyAssignment == nil {
+				continue
+			}
+
+			// We deepCopy the assignment, so if we delete unwanted domain, and
+			// There is no fit, we have the original assignment to retry with
+			assignment := psa.TopologyAssignment.DeepCopy()
+			nodeToReplace := wl.Annotations[kueuealpha.NodeToReplaceAnnotation]
+			tr.Count = deleteDomain(assignment, nodeToReplace)
+			if isStale, staleDomain := s.IsTopologyAssignmentStale(assignment); isStale {
+				result[tr.PodSet.Name] = tasPodSetAssignmentResult{TopologyAssignment: psa.TopologyAssignment, FailureReason: fmt.Sprintf("Domain %v not sound in the cache", staleDomain)}
+				return result
+			}
+			requiredDomain := s.requiredDomain(&tr, wl, assignment)
+			patchAssignment, reason := s.findTopologyAssignment(tr, assumedUsage, simulateEmpty, requiredDomain)
+			if patchAssignment != nil {
+				assignment = s.mergeTopologyAssignments(patchAssignment, assignment)
+			}
+			result[tr.PodSet.Name] = tasPodSetAssignmentResult{TopologyAssignment: assignment, FailureReason: reason}
+			if reason != "" {
+				return result
+			}
+			for _, domain := range patchAssignment.Domains {
+				domainID := utiltas.DomainID(domain.Values)
+				if assumedUsage[domainID] == nil {
+					assumedUsage[domainID] = resources.Requests{}
+				}
+				assumedUsage[domainID].Add(tr.SinglePodRequests.ScaledUp(int64(domain.Count)))
+			}
```

Comment:
This code looks very similar to the pre-existing code here: https://github.com/kubernetes-sigs/kueue/pull/5287/files?diff=unified&w=1#diff-5414dcb324187d1d853329b419691cde79632a6d77e1a7c2b503dc1a3688c081R428-R434, and it fixes the issue with assumedUsage for multiple domains. 

OTOH, it would probably be better to fix the issue in a separate PR so that we can cherry-pick easily. Once the issue is fixed we can commonize the code in a follow up.

Reply:
I'll revert the fix so we can cherry pick

Reply:
It's looks similar but there are actually a lot of changes so it's tricky to commonize those two branches, and I believe with the clear distinction between the preexisting logic and the new replacement logic it has better readability